pkg/agent/controller/networkpolicy/packetin.go

```diff
@@ -0,0 +1,184 @@
+	pktLength   uint16 // packet length of packetin
+	protocolStr string // protocol of the traffic logged
+}
+
+// initLogger is called while newing Antrea network policy agent controller.
+// Customize AntreaPolicyLogger specifically for Antrea Policies audit logging.
+func initLogger() error {
+	// logging file should be /var/log/antrea/networkpolicy/np.log
+	if _, err := os.Stat(logDir); os.IsNotExist(err) {
+		os.Mkdir(logDir, 0755)
+	}
+	file, err := os.OpenFile(logDir+logfileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0666)
+	if err != nil {
+		klog.Errorf("Failed to initialize logger to audit Antrea Policies %v", err)
+		return err
+	}
+
+	AntreaPolicyLogger = log.New(file, "", log.Ldate|log.Lmicroseconds)
+	// Use lumberjack log file rotation
+	AntreaPolicyLogger.SetOutput(&lumberjack.Logger{
+		Filename:   logDir + logfileName,
+		MaxSize:    500,  // allow max 500 megabytes for one log file
+		MaxBackups: 3,    // allow max 3 old log file backups
+		MaxAge:     28,   // allow max 28 days maintenance of old log files
+		Compress:   true, // compress the old log files for backup
+	})
+	klog.V(2).Info("Initialized Antrea-native Policy Logger for audit logging")
+	return nil
+}
+
+// HandlePacketIn is the packetin handler registered to openflow by Antrea network policy agent controller.
+// Retrieves information from openflow reg, controller cache, packetin packet to log.
+func (c *Controller) HandlePacketIn(pktIn *ofctrl.PacketIn) error {
+	if pktIn == nil {
+		return errors.New("empty packetin for Antrea Policy")
+	}
+
+	ob := new(logInfo)
+
+	// Get Network Policy log info
+	err := getNetworkPolicyInfo(pktIn, c, ob)
+	if err != nil {
+		return fmt.Errorf("received error while retrieving NetworkPolicy info: %v", err)
+	}
+
+	// Get packet log info
+	err = getPacketInfo(pktIn, ob)
+	if err != nil {
+		return fmt.Errorf("received error while handling packetin for NetworkPolicy: %v", err)
+	}
+
+	// Store log file
+	AntreaPolicyLogger.Printf("%s %s %s %s SRC: %s DEST: %s %d %s", ob.tableName, ob.npRef, ob.disposition, ob.ofPriority, ob.srcIP, ob.destIP, ob.pktLength, ob.protocolStr)
+	return nil
+}
+
+// getMatchRegField returns match to the regNum register.
+func getMatchRegField(matchers *ofctrl.Matchers, regNum uint32) *ofctrl.MatchField {
+	return matchers.GetMatchByName(fmt.Sprintf("NXM_NX_REG%d", regNum))
+}
+
+// getMatch receives ofctrl matchers and table id, match field.
+// Modifies match field to Ingress/Egress register based on tableID.
+func getMatch(matchers *ofctrl.Matchers, tableID binding.TableIDType) *ofctrl.MatchField {
+	// Get match from ingress/egress reg
+	for _, table := range append(openflow.GetAntreaPolicyEgressTables(), openflow.EgressRuleTable, openflow.EgressDefaultTable) {
+		if tableID == table {
+			return getMatchRegField(matchers, uint32(openflow.EgressReg))
+		}
+	}
+	for _, table := range append(openflow.GetAntreaPolicyIngressTables(), openflow.IngressRuleTable, openflow.IngressDefaultTable) {
```

Comment:
👀 My concern was that the set is more specific, this tableID is also used to determine if it is a `packetin` for log, e.g. traceflow has `packetin` from table `L2ForwardingOutTable`, which should not be further parsed.
But the comparison does make sense, not sure of the tradeoff here.

Reply:
Yes, a set can work as well, I just meant the current way that iterates all policy tables to determine its direction (EgressReg or IngressReg) looks not worth, but this might be minor improvement and can be done in future.